text/0000-render-element-modifiers.md

```diff
@@ -0,0 +1,570 @@
+we believe it is out of scope for this RFC. Default binding behavior could be
+added in the future, if a context API is decided on.
+
+## How we teach this
+
+Element modifiers will be new to everyone, so we're starting with a mostly blank
+slate. The only modifier that exists in classic Ember is `{{action}}`, and while
+most existing users will be familiar with it, that familiarity may not translate
+to the more general idea of modifiers.
+
+The first thing we should focus on is teaching _modifiers in general_. Modifiers
+should be seen as the place for any logic which needs to act directly on an
+element, or when an element is added to or removed from the DOM. Modifiers can
+be fully independent (for instance, a `scroll-to` modifier that transparently
+manages the scroll position of the element) or they can interact with the
+component (like the `did-render` and `will-destroy` modifiers). In all cases
+though, they are _tied to the render lifecycle of the element_, and they
+generally contain _side-effects_ (though these may be transparent and
+declarative, as in the case of `{{action}}` or the theoretical `{{scroll-to}}`).
+
+Second, we should teach the render modifiers specifically. We can do this by
+illustrating common use cases which can currently be solved with render hooks,
+and comparing them to using modifiers for the same solution.
+
+One thing we should definitely avoid teaching except in advanced cases is the
+_ordering_ of element modifiers. Ideally, element modifiers should be
+commutative, and order should not be something users have to think about. When
+custom element modifiers become widely available, this should be considered best
+practice.
+
+### Example: Scrolling an element to a position
+
+This sets the scroll position of an element, and updates it whenever the scroll
+position changes.
+
+Before:
+
+```hbs
+{{yield}}
+```
+```js
+export default Component.extend({
+  classNames: ['scroll-container'],
+
+  didRender() {
+    this.element.scrollTop = this.scrollPosition;
+  }
+});
+```
+
+After:
+
+```hbs
+<div {{did-render this.setScrollPosition @scrollPosition}} class="scroll-container">
+  {{yield}}
+</div>
+```
+```js
+export default class Component.extend({
+  setScrollPosition(element, scrollPosition) {
+    element.scrollTop = scrollPosition;
+  }
+})
+```
+
+#### Example: Adding a class to an element after render for CSS animations
+
+This adds a CSS class to an alert element in a conditional whenever it renders
+to fade it in, which is a bit of an extra hoop.
+
+Before:
+
+```hbs
+{{#if shouldShow}}
+  <div class="alert">
+    {{yield}}
+  </div>
+{{/if}}
+```
+```js
+export default Component.extend({
+  didRender() {
+    let alert = this.element.querySelector('.alert');
+
+    if (alert) {
+      alert.classList.add('fade-in');
+    }
+  }
+});
+```
+
+After:
+
+```hbs
+{{#if shouldShow}}
+  <div {{did-render this.fadeIn}} class="alert">
```

Comment:
I don't fully grok this example, how is this better than:

```hbs
<div class="fade-in"></div>
```

I am probably missing something...

Reply:
You generally need to append an element, _then_ add the CSS class to begin a transition. I'm not sure if that actually has to be after first paint of the element, or just when it is appended.